armv7a9-zynq7000/serial.c

```diff
@@ -0,0 +1,339 @@
+		l = min(serial->rxHead - serial->rxTail, len);
+	else
+		l = min(BUFFER_SIZE - serial->rxTail, len);
+
+	low_memcpy(buff, &serial->rxBuff[serial->rxTail], l);
+	cnt = l;
+	if ((len > l) && (serial->rxHead < serial->rxTail)) {
+		low_memcpy(buff + l, &serial->rxBuff[0], min(len - l, serial->rxHead));
+		cnt += min(len - l, serial->rxHead);
+	}
+	serial->rxTail = ((serial->rxTail + cnt) % BUFFER_SIZE);
+
+	low_sti();
+
+	return cnt;
+}
+
+
+int serial_write(unsigned int pn, const u8 *buff, u16 len)
+{
+	int l, cnt = 0;
+	serial_t *serial;
+
+	if (pn >= UARTS_MAX_CNT)
+		return ERR_ARG;
+
+	serial = &serial_common.serials[pn];
+
+	while (serial->txHead == serial->txTail && serial->tFull)
+		;
+
+	low_cli();
+	if (serial->txHead > serial->txTail)
+		l = min(serial->txHead - serial->txTail, len);
+	else
+		l = min(BUFFER_SIZE - serial->txTail, len);
+
+	low_memcpy(&serial->txBuff[serial->txTail], buff, l);
+	cnt = l;
+	if ((len > l) && (serial->txTail >= serial->txHead)) {
+		low_memcpy(serial->txBuff, buff + l, min(len - l, serial->txHead));
+		cnt += min(len - l, serial->txHead);
+	}
+
+	/* Initialize sending */
+	if (serial->txTail == serial->txHead)
+		*(serial->base + fifo) = serial->txBuff[serial->txHead];
+
+	serial->txTail = ((serial->txTail + cnt) % BUFFER_SIZE);
+
+	if (serial->txTail == serial->txHead)
+		serial->tFull = 1;
+
+	/* Enable TX FIFO empty irq */
+	 *(serial->base + ier) |= 1 << 3;
+	low_sti();
+
+	return cnt;
+}
+
+
+int serial_safewrite(unsigned int pn, const u8 *buff, u16 len)
+{
+	int l;
+
+	for (l = 0; len;) {
+		if ((l = serial_write(pn, buff, len)) < 0)
+			return ERR_MSG_IO;
+		buff += l;
+		len -= l;
+	}
+
+	return 0;
+}
+
+
+int serial_rxEmpty(unsigned int pn)
+{
+	serial_t *serial;
+
+	if (pn >= UARTS_MAX_CNT)
+		return ERR_ARG;
+
+	serial = &serial_common.serials[pn];
+
+	return serial->rxHead == serial->rxTail;
+}
+
+
+/* According to TRM:
+*  baud_rate = ref_clk / (bgen * (bdiv + 1))
+*  bgen: 2 - 65535
+*  bdiv: 4 - 255                             */
+static int serial_calcBaudarate(int pn, int baudrate)
+{
+	u32 bestDiff, diff;
+	u32 calcBaudrate;
+	u32 bdiv, bgen, bestBdiv = 4, bestBgen = 2;
+
+	serial_t *serial;
+
+	if (pn >= UARTS_MAX_CNT)
+		return ERR_ARG;
+
+	serial = &serial_common.serials[pn];
+
+	bestDiff = (u32)baudrate;
+
+	for (bdiv = 4; bdiv < 255; bdiv++) {
+		bgen = UART_REF_CLK / (baudrate * (bdiv + 1));
+
+		if (bgen < 2 || bgen > 65535)
+			continue;
+
+		calcBaudrate = UART_REF_CLK / (bgen * (bdiv + 1));
+
+		if (calcBaudrate > baudrate)
+			diff = calcBaudrate - baudrate;
+		else
+			diff = baudrate - calcBaudrate;
+
+		if (diff < bestDiff) {
+			bestDiff = diff;
+			bestBdiv = bdiv;
+			bestBgen = bgen;
+		}
+	}
+
+	*(serial->base + baudgen) = bestBgen;
+	*(serial->base + baud_rate_divider_reg0) = bestBdiv;
+
+	return ERR_NONE;
+}
+
+
+void serial_init(u32 baud, u32 *st)
+{
+	int i;
+	serial_t *serial;
+
+	static const struct {
+		volatile u32 *base;
+		u16 irq;
+	} info[] = {
```

Comment:
```suggestion
	} info[UARTS_MAX_CNT] = {
```
Array size should be equal.